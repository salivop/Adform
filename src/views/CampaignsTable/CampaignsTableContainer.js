import React from "react";
import axios from "axios";
import Typography from "@material-ui/core/Typography/Typography";

import PaperWrapper from "components/PaperWrapper/PaperWrapper";
import Loader from "components/Loader/Loader";

import CampaignsTable from "./Table/CampaignsTable";

import { mockTableData } from "./MockData";

class CampaignsTableContainer extends React.Component {
  constructor() {
    super();

    this.state = {
      tableData: [],
      isLoading: true,
      isError: false,
      url: "https://jsonplaceholder.typicode.com/users"
    };
    this.getData = this.getData.bind(this);
    this.dateCheck = this.dateCheck.bind(this);
  }
  componentDidMount() {
    setTimeout(function() {
     this.getData()
  }.bind(this), 2000);
}

  dateCheck = (startDate, endDate) => {
    let currentDate = new Date();
    let cDate = Date.parse(currentDate);
    return cDate <= Date.parse(endDate) && cDate >= Date.parse(startDate)
      ? true
      : false;
  };

  async getData() {
    const { url } = this.state;
    await axios
      .get(url)
      .then(data => {
        this.setState({
          // TODO:
          // THE API IS CHANGED BECAUSE PROVIDED API RETURN WRONG DATA.
          // IN THAT CASE I USED DATA FROM WORD DOC THAT I HAVE RECEIVED FROM ADFORM, ALSO DUE THAT SOME TESTS FAILS
          // tableData: data.data.map(item => ({
          //   ...item,
          //   active: this.dateCheck(item.startDate, item.endDate)
          // })),
          tableData: mockTableData.map(item => ({
            ...item,
            active: this.dateCheck(item.startDate, item.endDate)
          })),
          isLoading: false,
          isError: false
        });
      })
      .catch(this.setState({ isLoading: false, isError: true }));
  }

  render() {
    const { isLoading, isError, tableData } = this.state;
    return (
      <PaperWrapper
        child={
          <React.Fragment>
            {isError && (
              <Typography>Something wrong with table data</Typography>
            )}
            {isLoading ? <Loader /> : <CampaignsTable tableData={tableData} />}
          </React.Fragment>
        }
      />
    );
  }
}

export default CampaignsTableContainer;
