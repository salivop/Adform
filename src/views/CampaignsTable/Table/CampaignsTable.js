import React from "react";
import PropTypes from "prop-types";

import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";
import TableHead from "@material-ui/core/TableHead";
import TableSortLabel from "@material-ui/core/TableSortLabel";
import moment from "moment";

import { tableHeaders } from "./constants";

import { styles } from "./CampaignsTable.styles";

const desc = (a, b, orderBy) => {
  let orderByAType = moment(a[orderBy], "MM/DD/YYYY").isValid()
    ? Date.parse(a[orderBy])
    : a[orderBy];
  let orderByBType = moment(b[orderBy], "MM/DD/YYYY").isValid()
    ? Date.parse(b[orderBy])
    : b[orderBy];
  if (orderByBType < orderByAType) {
    return -1;
  }
  if (orderByBType > orderByAType) {
    return 1;
  }
  return 0;
};

const stableSort = (array, cmp) => {
  const stabilizedThis = array.map((el, index) => [el, index]);
  stabilizedThis.sort((a, b) => {
    const order = cmp(a[0], b[0]);
    if (order !== 0) return order;
    return a[1] - b[1];
  });
  return stabilizedThis.map(el => el[0]);
};

const getSorting = (order, orderBy) =>
  order === "desc"
    ? (a, b) => desc(a, b, orderBy)
    : (a, b) => -desc(a, b, orderBy);

const formatCurrencyValue = (value, currencyCode) =>
  parseFloat(value).toLocaleString(currencyCode, {
    style: "currency",
    currency: currencyCode,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2
  });

class CampaignsTable extends React.Component {
  constructor() {
    super();

    this.state = {
      order: "asc",
      orderBy: "id"
    };
    this.handleRequestSort = this.handleRequestSort.bind(this);
  }

  handleRequestSort = (event, property) => {
    const { order, orderBy } = this.state;
    const isDesc = orderBy === property && order === "desc";
    this.setState(prevState => ({
      order: isDesc ? "asc" : "desc",
      orderBy: property
    }));
  };
  render() {
    const { order, orderBy } = this.state;
    const { tableData } = this.props;

    const tableHeader = (order, orderBy, handleRequestSort) => (
      <TableHead>
        <TableRow>
          {tableHeaders.map(header => (
            <TableCell
              key={header.id}
              padding="default"
              sortDirection={orderBy === header.id ? order : false}
            >
              <TableSortLabel
                active={orderBy === header.id}
                direction={order}
                onClick={event => handleRequestSort(event, header.id)}
              >
                {header.label}
              </TableSortLabel>
            </TableCell>
          ))}
        </TableRow>
      </TableHead>
    );
    return (
      <React.Fragment>
        <Table>
          {tableHeader(order, orderBy, this.handleRequestSort)}
          <TableBody>
            {stableSort(tableData, getSorting(order, orderBy)).map(
              data =>
                Date.parse(data.endDate) >= Date.parse(data.startDate) && (
                  <TableRow hover key={data.id}>
                    <TableCell padding="none">{`Campaigns ${data.id}`}</TableCell>
                    <TableCell>
                      {data.userId ? data.name : "Unknown User"}
                    </TableCell>
                    <TableCell>{data.startDate}</TableCell>
                    <TableCell>{data.endDate}</TableCell>
                    <TableCell padding="none">
                      <React.Fragment>
                        <span style={styles.colorByDate(data.active)} />
                        {data.active ? "Active" : "Inactive"}
                      </React.Fragment>
                    </TableCell>
                    <TableCell>
                      {formatCurrencyValue(data.Budget, "USD")}
                    </TableCell>
                  </TableRow>
                )
            )}
          </TableBody>
        </Table>
      </React.Fragment>
    );
  }
}

CampaignsTable.propTypes = {
  tableData: PropTypes.array.isRequired
};

export default CampaignsTable;
