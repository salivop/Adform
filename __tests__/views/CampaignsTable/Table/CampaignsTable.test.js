import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";

import CampaignsTable from "src/views/CampaignsTable/Table/CampaignsTable";

describe("CampaignsTable", () => {
  let wrapper;

  beforeEach(() => (wrapper = shallow(<CampaignsTable tableData={[]} />)));

  it("should render correctly", () => {
    const component = wrapper;
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should start with an empty initial values", () => {
    expect(wrapper.state().order).toEqual("asc");
    expect(wrapper.state().orderBy).toEqual("id");
  });

  it("should have DOM elements", () => {
    expect(wrapper.find("Fragment").length).toEqual(1);
    expect(wrapper.find("WithStyles(ForwardRef(Table))").length).toEqual(1);
    expect(wrapper.find("WithStyles(ForwardRef(TableHead))").length).toEqual(1);
    expect(wrapper.find("WithStyles(ForwardRef(TableRow))").length).toEqual(1);
    expect(wrapper.find("WithStyles(ForwardRef(TableBody))").length).toEqual(1);
    expect(wrapper.find("WithStyles(ForwardRef(TableCell))").length).toEqual(6);
    expect(
      wrapper.find("WithStyles(ForwardRef(TableSortLabel))").length
    ).toEqual(6);
  });
});