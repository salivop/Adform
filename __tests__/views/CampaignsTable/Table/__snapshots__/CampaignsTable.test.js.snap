// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignsTable should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Table))>
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))
          key="id"
          padding="default"
          sortDirection="asc"
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={true}
            direction="asc"
            onClick={[Function]}
          >
            Name
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          key="name"
          padding="default"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            User Name
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          key="startDate"
          padding="default"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Start Date
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          key="endDate"
          padding="default"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            End Date
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          key="active"
          padding="default"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Active
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))
          key="Budget"
          padding="default"
          sortDirection={false}
        >
          <WithStyles(ForwardRef(TableSortLabel))
            active={false}
            direction="asc"
            onClick={[Function]}
          >
            Budget
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody)) />
  </WithStyles(ForwardRef(Table))>
</Fragment>
`;
